File: models/goods.go
package models

import (
	"database/sql"
)

type GoodsModel struct {
	db *sql.DB
}

func NewGoodsModel(db *sql.DB) *GoodsModel {
	return &GoodsModel{db}
}

func (g *GoodsModel) Find(where, opt map[string]interface{}) []*Good {
	var sql = "select * from goods"
	r, err := Find(sql, where, opt, g.Query)
	if err != nil {
		panic(err)
	}
	return r.([]*Good)
}

func (g *GoodsModel) Add(data map[string]interface{}) *DMLResult {
	var r = ExportKeyValues(data)
	var sql = MakeInsertSql("goods", r.Keys)
	var ret, err = DML(g.db, sql, r.Values...)
	if err != nil {
		panic(err)
	}
	return ret
}

func (g *GoodsModel) Count(where map[string]interface{}) int {
	var rows *sql.Rows
	var err error
	var sql = "select count(*) from goods"

	if where != nil {
		var r = BuildWhere(where)
		rows, err = g.db.Query(sql+r.Where, r.Args...)
	} else {
		rows, err = g.db.Query(sql)
	}
	if err != nil {
		panic(err)
	}
	defer rows.Close()
	var ret = 0
	for rows.Next() {
		var err = rows.Scan(&ret)
		if err != nil {
			panic(err)
		}
		break
	}
	err = rows.Err()
	if err != nil {
		panic(err)
	}
	return ret
}

type Good struct {
	Id           int
	Oid          string
	Goods_name   string
	Cat_id       string
	Shop_price   string
	Goods_img    string
	Goods_desc   string
	Goods_number int
	Is_best      int
	Is_new       int
	Is_hot       int
	Is_on_sale   int
	Created_at   string
	Updated_at   string
}

func (g *GoodsModel) Query(sql string, params ...interface{}) (interface{}, error) {
	rows, err := g.db.Query(sql, params...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var ret = make([]*Good, 0)
	for rows.Next() {
		var good = &Good{}
		var err = rows.Scan(&good.Id, &good.Oid, &good.Goods_name, &good.Cat_id, &good.Shop_price, &good.Goods_img, &good.Goods_desc,
			&good.Goods_number, &good.Is_best, &good.Is_new, &good.Is_hot, &good.Is_on_sale, &good.Created_at, &good.Updated_at)
		if err != nil {
			return nil, err
		}
		ret = append(ret, good)
	}
	err = rows.Err()
	if err != nil {
		return nil, err
	}
	return ret, nil
}

File: models/goods.js
var types = require('types');
var mtypes = require('./lib/types.js');
var db = require('./models/db.js');
var dbUtils = require('./models/dbUtils.js');
var utils = require('./lib/utils.js');

var db = db.getConn();


exports.count = function count(where) {
  var ret = null;
  if (where) {
    var r = dbUtils.exportKeyValues(where);
    var fields = r.keys;
    var values = r.values;
    var args = [];
    var arr = [];
    for (var i = 0; i < fields.length; i++) {
      var k = fields[i];
      var v = values[i];
      if (mtypes.isObject(v)) {
        if (v.$in) {
          var ins = v.$in;
          arr.push(k + ' in (' + dbUtils.makePlaceStr(ins.length) + ')');
          args = args.concat(ins);
        }
        continue;
      }
      arr.push(k + '=?');
      args.push(v);
    }
    var sql = 'select count(*) from goods where ' + arr.join(' and ');
    args.unshift(sql); // 将sql插入到最前面
    console.log(JSON.stringify(args));
    ret = db.query.apply(db, args);
  } else {
    ret = db.query('select count(*) from goods');
  }
  if (ret.err) {
    throw ret.err;
  }
  var rows = ret.value;
  while (rows.next()) {
    var n = types.newInt();
    var err = rows.scan(n);
    if (err) {
      throw err;
    }
    ret = types.intValue(n);
    break;
  }
  var err = rows.err();
  if (err) {
    throw err;
  }
  err = rows.close();
  if (err) {
    throw err;
  }
  return ret;
}

exports.add = function add(data) {
  var r = dbUtils.exportKeyValues(data);
  var sql = dbUtils.makeInsertSql('goods', r.keys);
  return dbUtils.insert(db, sql, r.values);
}

exports.delByOid = function delByOid(oid) {
  return dbUtils.delete(db, 'delete from goods where oid=?', [oid]);
}

exports.getByOid = function getByOid(oid) {
  var gs = find({ oid: oid }, {});
  if (gs.length > 0) {
    return gs[0];
  }
  return null;
}

function find(where, opt) {
  var sql = 'select * from goods';
  return dbUtils.find(sql, where, opt, query);
}

function query(sql, params) {
  var ret = [];
  params.unshift(sql);
  console.log('%j', params)
  var r = db.query.apply(db, params);
  if (r.err) {
    throw r.err;
  }
  var rows = r.value;
  while (rows.next()) {
    var id = types.newInt();
    var oid = types.newString();
    var goods_name = types.newString();
    var cat_id = types.newString();
    var shop_price = types.newString();
    var goods_img = types.newString();
    var goods_desc = types.newString();
    var goods_number = types.newInt();
    var is_best = types.newInt();
    var is_new = types.newInt();
    var is_hot = types.newInt();
    var is_on_sale = types.newInt();
    var created_at = types.newString();
    var updated_at = types.newString();

    var err = rows.scan(id, oid, goods_name, cat_id, shop_price, goods_img, goods_desc, goods_number, is_best, is_new, is_hot, is_on_sale, created_at, updated_at);
    if (err) {
      rows.close();
      throw err;
    }

    ret.push({
      id: types.intValue(id),
      oid: types.stringValue(oid),
      goods_name: types.stringValue(goods_name),
      cat_id: types.stringValue(cat_id),
      shop_price: types.stringValue(shop_price),
      goods_img: types.stringValue(goods_img),
      goods_desc: types.stringValue(goods_desc),
      goods_number: types.intValue(goods_number),
      is_best: types.intValue(is_best),
      is_new: types.intValue(is_new),
      is_hot: types.intValue(is_hot),
      is_on_sale: types.intValue(is_on_sale),
      created_at: types.stringValue(created_at),
      updated_at: types.stringValue(updated_at),
    });
  }
  var err = rows.err();
  if (err) {
    rows.close();
    throw err;
  }
  err = rows.close();
  if (err) {
    throw err;
  }
  return ret;
}

function updateByOid(oid, data) {
  var r = dbUtils.exportKeyValues(data)
  var sql = dbUtils.makeUpdateSql('goods', r.keys, ['oid'])
  // 添加where条件值
  r.values.push(oid)
  return dbUtils.update(db, sql, r.values)
}

function decrGoodsNum(oid, num) {
  var sql = 'update goods set goods_number=goods_number-? where oid=? and goods_number>=?';
  var args = [num, oid, num];
  return dbUtils.update(db, sql, args);
}

exports.decrGoodsNum = decrGoodsNum;
exports.find = find;
exports.updateByOid = updateByOid;
